File: 25_Reverse_Nodes_in_k-Group/Solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return None
        if not head.next or k == 1:
            return head

        t = []
        cur = head
        while cur:
            t.append(cur)
            cur = cur.next
        n = len(t)

        for i in range(0,(n//k)*k,k):
            j = i+k-1
            if i >= k:
                t[i-k].next = t[j]
            while j>i:
                t[j].next = t[j-1]
                j -= 1
        
        if n%k == 0:
            if n-2*k>-1:
                t[n-2*k].next = t[n-1]
            t[n-k].next = None
        else:
            if (n//k)*k-k > -1:
                t[(n//k)*k-k].next = t[(n//k)*k]
        if k<=n:
            return t[k-1]
        else:
            return t[0]

head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(3)
head.next.next.next = ListNode(4)
# head.next.next.next.next = ListNode(5)
k = 2
s =Solution()
s.reverseKGroup(head,k)

File: 25_Reverse_Nodes_in_k-Group/Solution.cc
#include <iostream>
#include <sstream>

using namespace std;

struct ListNode{
    int val;
    ListNode* next;
    ListNode(): val(0), next(nullptr){}
    ListNode(int x): val(x), next(nullptr){}
    ListNode(int x, ListNode* node):val(x), next(node){}
};

class Solution{

    public:
          static ListNode* reverseKGroup(ListNode* head, int k) {
              if(!head || !head->next) return head;
              ListNode* preHead = new ListNode();
              ListNode* pre = preHead;

              ListNode* cur = head;
              while(cur){
                  int c = 0;
                  ListNode* tNode = cur;
                  while(c<k && tNode){ c+=1; tNode=tNode->next;}
                  if(c==k){
                      ListNode* headNode=nullptr;
                      ListNode* tailNode=cur;
                      ListNode* pNode = nullptr;
                      while(c>0){
                          if(c==1) headNode = cur;
                          ListNode* tNode = cur->next;
                          cur->next = pNode;
                          pNode = cur;
                          cur = tNode;
                          c-=1;
                      }
                      pre->next = headNode;
                      tailNode->next = cur;
                      pre = tailNode;
                  }else{
                      pre->next = cur;
                      cur = nullptr;
                  }
              }
              return preHead->next;
          }
};

int main(){
    string line;
    while(getline(cin, line)){
        int k;
        istringstream is(line);
        is>>k;
        ListNode* preHead = new ListNode();
        ListNode* pre = preHead;
        int n;
        while(is>>n){
            pre->next = new ListNode(n);
            pre = pre->next;
        }
        auto node = Solution::reverseKGroup(preHead->next, k);
        cout<<line<<endl;
        while(node){
            cout<<node->val<<" ";
            node = node->next;
        }
        cout<<endl<<endl;;
    }
    return 0;
}
